### PR TITLE
Drift detection

### PR DESCRIPTION
**Pull request by cube2222**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (non-breaking change that adds documentation)

## Related issues

> Fix [#1]()

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Pull Request is no longer marked as "draft"
- [ ] Reviewers have been assigned
- [ ] Changes have been reviewed by at least one other engineer

### CODE REVIEW

File: spacelift/resource_drift_detection.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+			Reconcile: graphql.Boolean(d.Get("reconcile").(bool)),
+			Schedule:  scheduleExpressions,
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "DriftDetectionUpdate", &mutation, variables); err != nil {
+		return diag.Errorf("could not create drift detection integration for the stack: %v", err)
+	}
+
+	if d.Id() == "" {
```
Comment (marcinwyszynski) [inline spacelift/resource_drift_detection.go:79]: That's always going to be true for the `create` phase.

File: spacelift/resource_drift_detection.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			Reconcile: graphql.Boolean(d.Get("reconcile").(bool)),
+			Schedule:  scheduleExpressions,
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "DriftDetectionCreate", &mutation, variables); err != nil {
+		return diag.Errorf("could not update drift detection integration for the stack: %v", err)
+	}
+
+	if d.Id() == "" {
```
Comment (marcinwyszynski) [inline spacelift/resource_drift_detection.go:113]: You're not supposed to do that in the `update`. The assumption is that the ID is set (and immutable) for a resource that's being updated.

File: spacelift/resource_drift_detection.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+	integration := query.Stack.Integrations.DriftDetection
+
+	d.Set("reconcile", integration.Reconcile)
+
+	schedule := make([]interface{}, len(integration.Schedule))
+	for i, expr := range integration.Schedule {
+		schedule[i] = expr
+	}
+	d.Set("schedule", schedule)
```
Comment (marcinwyszynski) [inline spacelift/resource_drift_detection.go:168]: I think Terraform SDK docs suggest that you should check for the error when calling `Set` with a collection type (list, set, map) argument.

File: spacelift/resource_drift_detection_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			scheduleStrs := make([]string, len(schedule))
+			for i := range schedule {
+				scheduleStrs[i] = `"` + schedule[i] + `"`
+			}
+
+			return fmt.Sprintf(`
+				resource "spacelift_stack" "test" {
+					branch     = "master"
+					repository = "demo"
+					name       = "Test stack %s"
```
Comment (marcinwyszynski) [inline spacelift/resource_drift_detection_test.go:30]: Technically this means that if you're unlucky you could trigger a job, no?

**Review by marcinwyszynski — APPROVED** [review]

File: spacelift/resource_drift_detection.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	variables := map[string]interface{}{
+		"stack": toID(stackID),
+		"input": structs.DriftDetectionIntegrationInput{
+			Reconcile: graphql.Boolean(d.Get("reconcile").(bool)),
+			Schedule:  scheduleExpressions,
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "DriftDetectionUpdate", &mutation, variables); err != nil {
+		return diag.Errorf("could not create drift detection integration for the stack: %v", err)
```
Comment (alexjurkiewicz) [inline spacelift/resource_drift_detection.go:76]: Terraform's error message will not include the stack name if this is shown, which will make things harder to debug especially when creating multiple drift configurations at once.

Please consider including the stack name in error messages about sub-stack resources. And in general showing the parent resource ID in child resource error messages. 🙏

**Pull request merged**